File: src/0x09강 - BFS/1926_그림.cpp
/*
    vector의 최댓값, 최솟값:
    #include<algorithm>
    *max_element(v.begin(), v.end())
    *max_element(v.begin(), v.end())
*/
#include<iostream>
#include<queue>
#include<vector>
#include<algorithm>
#define MAX 500
using namespace std;

int n, m;

int bfs(int x, int y, int cnt, int graph[MAX][MAX])
{
    queue<pair<int,int>> q;
    q.push({x, y});
    graph[x][y] = 0;
    cnt++;

    int dx[4] = {-1, 1, 0, 0};
    int dy[4] = {0, 0, -1, 1};

    while(!q.empty())
    {
        pair<int,int> loc = q.front();
        q.pop();

        for(int i=0; i<4; i++)
        {
            int nx = loc.first + dx[i];
            int ny = loc.second + dy[i];

            if(nx >= 0 && nx < n && ny >= 0 && ny < m)
            {
                if(graph[nx][ny]==1)
                {
                    q.push({nx, ny});
                    graph[nx][ny] = 0;
                    cnt++;
                }
            }
        }
    }

    return cnt;
}

int main()
{
    int graph[MAX][MAX];
    vector<int> ans;
    cin >> n >> m;
    for(int i=0; i<n; i++)
    {
        for(int j=0; j<m; j++)
        {
            cin >> graph[i][j];
        }
    }

    for(int i=0; i<n; i++)
    {
        for(int j=0; j<m; j++)
        {
            if(graph[i][j] == 1)
            {
                ans.push_back(bfs(i,j,0,graph));
            }
        }
    }
    cout << ans.size() << endl;
    if(ans.empty())
    {
        cout << 0 << endl;
    }
    else
    {
        cout << *max_element(ans.begin(), ans.end()) << endl;
    }
}

File: src/0x09강 - BFS/1926_그림.py
'''
그림의 개수, 그림 중 가장 넓은 것의 넓이를 출력하라

[함정] 단, 그림이 하나도 없는 경우에는 가장 넓은 그림의 넓이는 0이다.
'''
from collections import deque

n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]


def bfs(x, y, cnt, graph):
    q = deque([(x, y)])
    graph[x][y] = 0
    cnt += 1

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                if graph[nx][ny] == 1:
                    q.append((nx, ny))
                    graph[nx][ny] = 0
                    cnt += 1

    return cnt


ans = []
for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            ans.append(bfs(i, j, 0, graph))

print(len(ans))
if len(ans) == 0:
    print(0)
else:
    print(max(ans))
